src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs

```diff
@@ -22,43 +22,129 @@ public void SslStream_ClientSendsSNIServerReceives_Ok(string hostName)
+            {
+                Assert.True(false, "LocalCertificateSelectionCallback called when AuthenticateAsServerAsync was expected to fail.");
+                return null;
+            });
+
+            var validationCallback = new RemoteCertificateValidationCallback((object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors) =>
+            {
+                Assert.Equal(serverCert, certificate);
+                return true; 
+            });
+
+            VirtualNetwork vn = new VirtualNetwork();
+            using (VirtualNetworkStream serverStream = new VirtualNetworkStream(vn, isServer: true),
+                                        clientStream = new VirtualNetworkStream(vn, isServer: false))
+            using (SslStream server = new SslStream(serverStream, false, null, selectionCallback),
+                             client = new SslStream(clientStream, leaveInnerStreamOpen: false, validationCallback))
             {
                 Task clientJob = Task.Run(() => {
                     client.AuthenticateAsClient(hostName);
+                    Assert.True(false, "RemoteCertificateValidationCallback called when AuthenticateAsServerAsync was expected to fail.");
                 });
 
                 SslServerAuthenticationOptions options = DefaultServerOptions();
-
-                int timesCallbackCalled = 0;
                 options.ServerCertificateSelectionCallback = (sender, actualHostName) =>
                 {
                     timesCallbackCalled++;
                     Assert.Equal(hostName, actualHostName);
                     return serverCert;
                 };
 
-                var cts = new CancellationTokenSource();
-                server.AuthenticateAsServerAsync(options, cts.Token).Wait();
+                await Assert.ThrowsAsync<InvalidOperationException>(() => server.AuthenticateAsServerAsync(options, CancellationToken.None));
 
-                Assert.Equal(1, timesCallbackCalled);
-                clientJob.Wait();
-            },
-            (object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors) =>
-                {
-                    Assert.Equal(serverCert, certificate);
-                    return true;
+                Assert.Equal(0, timesCallbackCalled);
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(HostNameData))]
+        public async void SslStream_ServerCallbackNotSet_UsesLocalCertificateSelection(string hostName)
+        {
+            X509Certificate serverCert = Configuration.Certificates.GetSelfSignedServerCertificate();
+
+            int timesCallbackCalled = 0;
+
+            var selectionCallback = new LocalCertificateSelectionCallback((object sender, string targetHost, X509CertificateCollection localCertificates, X509Certificate remoteCertificate, string[] issuers) =>
+            {
+                Assert.Equal(targetHost, string.Empty);
```

Comment:
expected parameters are listed first in these `Assert.Equal` methods (see similar code in file).  So this should be:

```c#
Assert.Equal(string.Empty, targetHost);
```

Reply:
Thanks. I'll check over the rest of these as well.